### DISCUSSION TITLE
"Prefetchable" Atoms

### DISCUSSION THREAD

**Discussion by dvd900** [pmndrs/jotai#2326] [Ideas]
I have a project where I'm using atoms as a primitive for api queries. I haven't found a good enough reason to use something like `tan-stack `. However, `atomsWithRefresh` had it's limitations for me. I wanted a way to have event based driven fetches of queries, in anticipation of it being needed later in the users experience.However it seemed like the only way to interact with `atomsWithRefresh` was to mount it using a `useAtom` call in a component, which would immediately fetch the data. I composed an atom to sit on top that would allow me to prefetch stuff.

```typescript
type AtomWithRefresh<T> = WritableAtom<Promise<T>, [boolean?], Promise<Promise<T>>>;
//This allows us to mount atoms to prefetch them without suspending
export function preFetchableAtomWithRefresh<T>(atomWithRefresh: AtomWithRefresh<T>) {
  const fetch = atom(false);
  const valueAtom = atom(
    (get) => {
      const shouldFetch = get(fetch);
      if (shouldFetch) {
        return get(unwrap(atomWithRefresh));
      } else {
        return null;
      }
    },
    (_, set) => {
      set(fetch, true);
    }
  );
  return valueAtom;
}
``` 

Does this make sense? 

An example usage could look something like 

```typescript
  const [, fetch] = useAtom(prefetchableAtom);
  const [prefetch, setPrefetch] = useState(false)
  useEffect(() => {
    if (prefetch) {
      fetch();
    }
  }, [prefetch]);
```

**Comment by dai-shi** [comment]
atomWithRefresh is a kind of hack to trigger suspense without changing data (server may return different data though.)
If you don't need suspense, things must be much simpler.
For imperative fetch, you can have fetch logic in the atom's write function.

Sorry, if I misunderstood your question. Feel free to correct me.

**Reply by dvd900** [reply]
I want to prefetch data like above in a HOC at certain points in my app without halting the UI or suspending, so while I prefetch I want to avoid suspense. However I can't guarantee that any of those preloads are complete when my lower order components that consume those atoms, which is where I need the refresh capabilities and where I want to suspend. 

`unwrap` gets me very close to a solution, i just wanted the ability to dispatch a load, rather than using a mounted component to kick of the fetch.

Maybe the architecture is strange or my use case is very particular, but I'm raising this to make sure I'm not missing anything obvious.

**Reply by dai-shi** [reply]
If you want to avoid "suspense" completely, you should fetch data outside React (like a callback, or useEffect), which is one of popular practice.

Otherwise (= needs suspense), I think these two pointers might help:
- https://github.com/jotaijs/jotai-suspense `usePrepareAtoms`
- https://blog.axlight.com/posts/how-to-use-jotai-and-use-transition-for-mutation/

(I'm still a bit confused with your architecture. maybe a small code snippet that doesn't "prefetch" might help.)
